src/parquet/file/writer-internal.cc

```diff
@@ -0,0 +1,260 @@
+  uint8_t* uncompressed_ptr = uncompressed_data->mutable_data();
+  memcpy(uncompressed_ptr, repetition_levels->data(), repetition_levels->size());
+  uncompressed_ptr += repetition_levels->size();
+  memcpy(uncompressed_ptr, definition_levels->data(), definition_levels->size());
+  uncompressed_ptr += definition_levels->size();
+  memcpy(uncompressed_ptr, values->data(), values->size());
+
+  // Compress the data
+  int64_t compressed_size = uncompressed_size;
+  std::shared_ptr<OwnedMutableBuffer> compressed_data = uncompressed_data;
+  if (compressor_) {
+    // TODO(PARQUET-592): Add support for compression
+    // int64_t max_compressed_size = compressor_->MaxCompressedLen(
+    // uncompressed_data.size(), uncompressed_data.data());
+    // OwnedMutableBuffer compressed_data(compressor_->MaxCompressedLen(
+    // uncompressed_data.size(), uncompressed_data.data()));
+  }
+  // Compressed data is not needed anymore, so immediately get rid of it.
+  uncompressed_data.reset();
+
+  format::DataPageHeader data_page_header;
+  data_page_header.__set_num_values(num_rows);
+  data_page_header.__set_encoding(ToThrift(encoding));
+  data_page_header.__set_definition_level_encoding(ToThrift(definition_level_encoding));
+  data_page_header.__set_repetition_level_encoding(ToThrift(repetition_level_encoding));
+  // TODO(PARQUET-593) statistics
+
+  format::PageHeader page_header;
+  page_header.__set_type(format::PageType::DATA_PAGE);
+  page_header.__set_uncompressed_page_size(uncompressed_size);
+  page_header.__set_compressed_page_size(compressed_size);
+  page_header.__set_data_page_header(data_page_header);
+  // TODO(PARQUET-594) crc checksum
+
+  int64_t start_pos = sink_->Tell();
+  SerializeThriftMsg(&page_header, sizeof(format::PageHeader), sink_);
+  int64_t header_size = sink_->Tell() - start_pos;
+  sink_->Write(compressed_data->data(), compressed_data->size());
+
+  metadata_->meta_data.total_uncompressed_size += uncompressed_size + header_size;
+  metadata_->meta_data.total_compressed_size += compressed_size + header_size;
+  metadata_->meta_data.num_values += num_values;
+
+  return sink_->Tell() - start_pos;
+}
+
+// ----------------------------------------------------------------------
+// RowGroupSerializer
+
+int RowGroupSerializer::num_columns() const {
+  return schema_->num_columns();
+}
+
+int64_t RowGroupSerializer::num_rows() const {
+  return num_rows_;
+}
+
+const SchemaDescriptor* RowGroupSerializer::schema() const  {
+  return schema_;
+}
+
+ColumnWriter* RowGroupSerializer::NextColumn() {
+  if (current_column_index_ == schema_->num_columns() - 1) {
+    throw ParquetException("All columns have already been written.");
+  }
+  current_column_index_++;
+
+  if (current_column_writer_) {
+    total_bytes_written_ += current_column_writer_->Close();
+  }
+
+  auto column_descr = schema_->Column(current_column_index_);
+  auto col_meta = &metadata_->columns[current_column_index_];
```

Comment:
Yeah, if the type is sufficiently clear from context (e.g. if there is a `static_cast<T>`), see https://google.github.io/styleguide/cppguide.html#auto.